스프링 자동차경주 리팩토링

givxq opened the pull request:
스프링 자동차경주 리팩토링

givxq committed:
[feature] SpringRacingCar
- racing car 도메인 추가

givxq committed:
[feature] SpringRacingCar
- racing car Controller 추가

givxq committed:
[feature] SpringRacingCar
- racing car Repository  추가

givxq committed:
[feature] SpringRacingCar
- README 추가

givxq committed:
[feature] SpringRacingCar
- PlayResult Repository 추가

givxq committed:
[feature] SpringRacingCar
- racing car Controller 수정

givxq committed:
[feature] SpringRacingCar
- racing car Model/Domain 수정

givxq committed:
[feature] SpringRacingCar
- README 수정

givxq committed:
[feature] SpringRacingCar
- READMEc- 체크리스트 수정

givxq committed:
[feature] SpringRacingCar
- 메서드 체이닝으로 가독성 증가

givxq committed:
[feature] SpringRacingCar
- Car 테이블 단수로 변경

givxq committed:
[feature] SpringRacingCar
- Car 객체 hashcode, equlas 메소드 추가

givxq committed:
[feature] SpringRacingCar
- gradle 설정 commit

givxq committed:
[feature] SpringRacingCar
- moveForward private로 수정

givxq committed:
[feature] SpringRacingCar
- move 메소드 오버라이딩

givxq committed:
[feature] SpringRacingCar
- CarRandom -> CarMoveUtil 클래스 이름 변경

givxq committed:
[feature] SpringRacingCar
- test 코드 rollback 옵션 제거

givxq committed:
[feature] SpringRacingCar
- data.sql 삭제

givxq committed:
[feature] SpringRacingCar
- application.yml 오타수정

givxq committed:
[feature] SpringRacingCar
- RacingCarView 클래스 삭제

givxq committed:
[feature] SpringRacingCar
- 사용하지 않는 메소드 제거

givxq committed:
[feature] SpringRacingCar
- 미션 요구사항을 위한 sercive 생성자 변경

- 웹 애플리케이션을 구현하기 위해 기존 콘솔 애플리케이션을 수정하지 않습니다.
- 필요하다면 콘솔 애플리케이션에서 코드를 가져와도 됩니다. 중복 제거를 위한 리팩터링은 2단계에서 진행합니다.

givxq committed:
[feature] SpringRacingCarToJPA
- Cars 객체 생성

givxq committed:
[feature] SpringRacingCarToJPA
- JPA / Service 변경

givxq committed:
[feature] SpringRacingCarToJPA
- 필요없는 class 삭제

givxq committed:
Merge pull request #1 from givxq/springRacingCarToJPA

jpa 개발 merge

givxq committed:
[feature] SpringRacingCar
- 미사용 로직 삭제

givxq committed:
[feature] SpringRacingCar
- console 출력 변경

givxq committed:
[feature] SpringRacingCar
- 중복로직 제거

givxq committed:
[feature] SpringRacingCar
- 미사용 import 제거

givxq committed:
[feature] SpringRacingCar
- 중복 소스 제거

givxq committed:
[feature] 2step_refactoring
- CommandLineRunner 소스 test코드와 분리

givxq committed:
[feature] 2step_refactoring
- 테스트 코드 작성 및 코드 리팩토링

givxq committed:
[feature] 2step_refactoring
- 미사용로직 삭제

givxq committed:
Merge branch 'givxq' into 2step_refactoring

seokhongkim commented on src/main/java/racingcar/RacingCarApplication.java:
```diff
@@ -1,31 +1,19 @@
 	RacingCarService racingCarService;
 
 	public static void main(String[] args) {
 		SpringApplication.run(RacingCarApplication.class, args);
```
인덴트 설정이 탭을 사용하도록 되어 있네요! :)

프로그래밍 요구 사항에 맞춰서 개발해보시면 좋겠습니다.
* 자바 코드 컨벤션을 지키면서 프로그래밍한다.
   * 기본적으로 [Google Java Style Guide](https://google.github.io/styleguide/javaguide.html)을 원칙으로 한다.
   * 단, 들여쓰기는 '2 spaces'가 아닌 '4 spaces'로 한다.

givxq replied:
ide를 재설치 했더니 설정이 변경되었네요.. 확인 감사합니다!

seokhongkim commented on src/main/java/racingcar/RacingCarApplication.java:
```diff
@@ -1,31 +1,19 @@
-public class RacingCarApplication implements CommandLineRunner {
+public class RacingCarApplication {
 	@Autowired
 	RacingCarService racingCarService;
```
접근 제어자를 default 로 사용하신 이유가 있을까요?

givxq replied:
실수로 빠졌네요..ㅎㅎ 추가했습니다👍

seokhongkim commented on src/test/java/racingcar/RacingCarApplicationTests.java:
```diff
@@ -27,6 +29,17 @@ class RacingCarApplicationTests {
 
+	@Transactional
+	@Test
+	public void makeCarsAndfindAllTest() {
```
같은 테스트가 2개 존재하네요! 하나는 삭제해도 될 것 같습니다. :)

seokhongkim commented on src/main/java/racingcar/domain/Car.java:
```diff
@@ -41,10 +41,20 @@ public Car() {
+        return this.position;
     }
 
     public void setPosition(int position) {
```
테스트에서만 사용되는 메서드인데 필요할까요?
테스트에서는 `move` 기능을 이용해서 위치를 변경하고 저장하는 것으로 검증을 해볼수도 있을 것 같습니다. :)

`move`가 항상 이동하는 것을 보장하기 위해서 `CarMoveUtil`에 static 으로 의존하는 관계를 동적으로 전달할 수 있도록 수정해보셔도 좋겠습니다.

givxq replied:
getPosition은 현재는 필요하지 않습니다.
과거 tdd 방법으로 구현시 테스트 코드를 위해 있던 부분이였습니다. 

기존 move(true) 함수 대신 CarMoveUtil을 사용하는게 좋을것같다는 말씀이시죠?
수정해보겠습니다.

givxq replied:
말씀하신대로 수정해보려했는데, 기존에도 move(),move(boolean) 함수가 있어서 CarMoveUtil에 추가하는게 어떤 이점이 있는지 잘 모르겠습니다.

givxq replied:
그리고 getPosition의 경우는 cars 객체에서 maxDistance를 가져오기 위해 필요한 상태네요.

seokhongkim replied:
`getPosition`은 사용해도 될 것 같고, `setPosition`은 제거해도 되지 않을까요?

현재 `move()`, `move(boolean)`을 모두 제공하는 이유는 프로덕션 코드에서 필요하기 때문일까요? 테스트를 위해서 메서드가 추가로 제공되고 있는 것은 아닐까요?
예전 [자동차 경주 미션 2단계](https://edu.nextstep.camp/s/jmQRkjTB/ls/XLkqdhac)에서 자동차 이동 조건이 난수에 의해 결정되기 때문에 테스트하기 어려운 것을, 난수 생성을 외부에서 주입하는 형태로 변경함으로써 테스트가 용이한 구조로 만들 수 있다는 점을 참고해보시면 좋겠습니다.

즉, `move(boolean)`만 제공을 하거나(외부에서 이동 여부를 결정하는 형태, 이 경우 boolean 값을 전달해야 할 이유가 있을까요? 이동할 때만 `move` 메서드를 호출하면 되는 것 아닐까요?), `move(numberGenerator)`와 같이 난수 생성기를 외부에서 메서드의 인자로 전달하거나, 자동차의 생성자에서 난수 생성기를 전달 받아서 필드로 관리한다면 어떨까요?

단순히 테스트 작성만 용이한 것을 떠나서, 행위를 임의로 변경할 수 있다는 것은 이동 조건에 대한 정책이 달라져도 `Car`는 재사용할 수 있는 코드가 된다는 것으로 생각되기도 하네요! :)

seokhongkim commented on src/main/java/racingcar/domain/Cars.java:
```diff
@@ -47,27 +56,31 @@ public void save() {
+    public List<String> getWinnerCars() {
         return cars.stream()
-                .filter(car -> car.getPosition() == maxDistance)
+                .filter(car -> car.getPosition() == getMaxDistance())
```
`getPosition`으로 값을 꺼내서 사용하지 말고, `car.isAtPosition(getMaxDistance())`와 같이 car 에게 해당 위치에 있는지 물어보는 형태로 구현해보셔도 좋겠습니다.

givxq replied:
좋은 의견 감사합니다.👍

seokhongkim commented on src/main/java/racingcar/domain/Cars.java:
```diff
@@ -7,18 +7,22 @@
 import racingcar.utils.RacingCarUtils;
 import racingcar.view.RacingResultView;
 
 @Service
```
이 클래스를 서비스로 관리하신 이유가 있을까요?

`List<Car>`를 필드로 가지고 있는데, 동시에 자동차 경주를 진행하는 요청이 오면 무슨 일이 발생할까요? 조금 더 확장하여, 스프링의 빈의 스코프에 무엇이 있고, 기본 스코프는 무엇이며, 해당 빈이 상태를 가지고 있으면 어떤 문제가 발생할 수 있을까요?

givxq replied:
이부분이 말씀하신대로 service로 등록하면 안될것같네요👍

seokhongkim commented on src/main/java/racingcar/domain/Cars.java:
```diff
@@ -47,27 +56,31 @@ public void save() {
     }
 
     public void printResult() {
         RacingResultView.printResult(getWinnerNames(), cars);
```
1. 도메인이 뷰에 의존하면 어떤 문제가 발생할 수 있을까요?
2. 이 기능은 콘솔 어플리케이션에서만 필요한 것으로 보입니다. 콘솔 어플리케이션에서만 사용하도록 이동해보면 어떨까요?

givxq replied:
말씀하신대로 수정하는편이 좋을것같네요 👍

seokhongkim commented on src/main/java/racingcar/service/RacingCarService.java:
```diff
@@ -26,4 +31,16 @@ public RacingResponse startRacing(String names, int targetDistance) {
+    public List<RacingResponse> getRacingHistory() {
+        List<RacingResponse> racingHistory = new ArrayList<>();
+
+        playResultRepository.findAll().forEach(playResult -> {
```
타입을 변환할 때는 `stream`의 `map` API를 사용해보는 것은 어떨까요?

givxq replied:
map으로 변경해보았습니다👍

seokhongkim commented on src/test/java/racingcar/RacingCarApplicationTests.java:
```diff
@@ -118,4 +130,62 @@ public void findAllPlayResultTest() {
+	}
+
+	@Test
+	public void makeCarTest() {
```
`Car`, `Cars` 클래스에 대한 테스트는 각 클래스에 대한 단위 테스트 파일로 분리해보면 어떨까요?

givxq replied:
넵 말씀하신대로 변경하는것이 좋을듯합니다. 수정해보겠습니다👍

seokhongkim commented on src/main/java/racingcar/service/RacingCarService.java:
```diff
@@ -15,6 +19,7 @@ public class RacingCarService {
     private final PlayResultRepository playResultRepository;
 
+    @Transactional
     public RacingResponse startRacing(String names, int targetDistance) {
```
미션의 요구사항을 봤을 때, 콘솔 애플리케이션은 구현의 변경이 없어야 할 것으로 보입니다. :)
공통 로직을 추출하되, 첫 번째 미션이었던 자동차 경주와 동일하게 데이터는 저장하지 않고 경주의 결과만 반환되어야 할 것으로 보입니다.

### 리팩터링 - 중복 코드 제거
* console application과 web application의 중복 코드를 제거합니다.
   * 두 application은 입출력과 데이터 저장 방식을 제외하고는 내부 비즈니스 로직은 동일해졌습니다.
   * 두 application의 비즈니스 로직은 XXXService라는 객체를 도출 하여 중복 제거를 할 수 있습니다. (레이어드 아키텍처 참고)

givxq replied:
제가 공통에 신경쓰다보니 프로그램을 합쳐버렸네요.....ㅎㅎ
말씀해주신 내용과 코멘트들 참고하여 다시 리뷰요청드리겠습니다.
월요일 이른 오전부터 리뷰 감사합니다. 월요팅하세요!

seokhongkim reviewed (changes requested):
진범님, 2단계 미션 잘 구현하셨습니다.
미션 요구 사항과 조금 다른 것으로 보이는 부분이 있어 해당 내용에 대한 코멘트와 함께 코드에 대한 코멘트를 조금 남겼습니다.
코멘트 확인 부탁드립니다!

givxq committed:
[feature] 2step_refactoring
- comment 내용수정

seokhongkim commented on src/main/java/racingcar/service/RacingCarService.java:
```diff
@@ -10,36 +10,36 @@
 
         playResult.setWinners(cars.getWinnerNames());
         playResultRepository.save(playResult);
         cars.save();
```
레포지토리에게 `Cars`를 전달해서 내부에 있는 `cars` 리스트를 저장하도록 변경해보는 건 어떨까요?
현재는 자동차 목록을 관리하는 컬렉션이 저장소까지 관리하는 역할을 겸하고 있는 것으로 보입니다. :)

givxq replied:
좋은 의견 감사합니다.
car 내부에서 생성할 수 있도록 CarsRepository를 추가하였습니다.

console, web 각각  JpaRepository 를 상속 받는 save 를 구현하였고,
외부에서 주입하여 로직을 재사용할수 있도록 개발하였습니다.

seokhongkim commented on src/main/java/racingcar/service/RacingCliCarService.java:
```diff
@@ -0,0 +1,15 @@
+import racingcar.view.RacingResultView;
+
+public class RacingCliCarService {
+    public void startRacing(String names, int targetDistance) {
```
미션의 내용 중 아래와 같은 내용이 있습니다.

### 리팩터링 - 중복 코드 제거

* console application과 web application의 중복 코드를 제거합니다.
   * 두 application은 입출력과 데이터 저장 방식을 제외하고는 내부 비즈니스 로직은 동일해졌습니다.
   * 두 application의 비즈니스 로직은 XXXService라는 객체를 도출 하여 중복 제거를 할 수 있습니다. (레이어드 아키텍처 참고)

--- 

`RacingCarService`를 웹, 콘솔 애플리케이션에서 함께 사용할 수 있는 방법은 없을까요?
동일한 로직이고 동일한 결과(우승자 목록과 이동 거리)를 반환하는데, 데이터를 중간에 저장하느냐의 차이만 있는 것 같습니다.

데이터를 저장하는 레포지토리만 개별적으로 외부에서 전달해보시면 좋겠습니다. 웹에서는 DB에 저장하는 레포지토리를 전달하고, 콘솔에서는 메모리에 보관하거나 아무것도 하지 않는 레포지토리를 전달함으로써 서비스를 재사용할 수 있을 것 같아요!

givxq replied:
좋은 의견 감사합니다.
말씀해주신 내용처럼 console, web 각각 JpaRepository를 custom 하여 구현하였고,
의존성을 외부에서 주입하여 재사용가능하도록 변경해보았습니다.

seokhongkim reviewed (changes requested):
진범님, 2단계 미션 코멘트 잘 반영하셨네요. 👍 
수정하신 코드를 보고 조금 더 고민해보시면 좋을 것 같은 내용이 있어서 코멘트 남겼습니다.
코멘트 확인 부탁드립니다.

givxq committed:
[feature] 2step_refactoring
- Cars Repository  추가
- CarConsoleRepository, CarWebRepository 분리 및 Service 재사용
- Cars 객체 정적 팩토리 메서드 추가

givxq commented:
CommandLineRunner 을 이용하여 console main을 만들었습니다.

이유는 console, web에서 사용할 부분을  Jpa  Repository를 커스터마이징하여 구현하다보니
console 프로그램에서도 스프링 bean을 사용하게 되었습니다.

 main을 2개 만드는 대신 CommandLineRunner run으로 콘솔을 만들게 된 점 참고부탁드립니다.🙏🏻

givxq committed:
[feature] 2step_refactoring
- CarsConsoleRepository -> CarConsole 이름변경

seokhongkim reviewed (changes requested):
진범님, 2단계 미션 코멘트 반영하신 것 확인 완료했습니다. :)
추가로 수정할 수 있는 부분이 있는데, 이게 필수는 아닌 것으로 보여 진범님의 의사 여쭤보고 머지하려고 합니다.

콘솔 어플리케이션이 반드시 스프링을 사용해야 할까요? 예를 들어, 아래와 같이 콘솔 어플리케이션의 메인을 구현할 수도 있습니다.
```java
public class RaingCarCliApplication {

    public static void main(String[] args) {
        CarsRepository carsRepository = new CarsRepository(null) {
            @Override
            public void save(Cars cars) {
                // do nothing.
            }
        };
        PlayResultRepository playResultRepository = new PlayResultRepository() {
            // PlayResultRepository 를 구현하면 JPARepository를 구현해야 해서 많은 메서드의 구현이 필요하므로, PlayResultRepository를 감싼 다른 구현 계층을 추가하는 방법이 더 좋을 것 같습니다. :)
            @Override
            public PlayResult save(PlayResult playResult) {
                return null;
            }
        };
        RacingCarService racingCarService = new RacingCarService(carsRepository, playResultRepository);

        while (true) {
            String carNames = RacingInputCarView.getCarNames();
            int targetDistance = RacingInputCarView.getTargetDistance();
            RacingResponse racingResponse = racingCarService.startRacing(carNames, targetDistance);
            RacingResultView.printRacingResponse(racingResponse);
        }
    }
}
```
이렇게 하면 스프링의 의존성 주입을 사용하지 않으면서도, 서비스와 도메인 로직을 그대로 재사용할 수 있습니다. :)

직접 수정해보고 싶으시면 반영 부탁드립니다.
이미 다양한 방법으로 구현해보셨기에 충분하다고 생각하시면, 코드 변경 없이 리뷰 요청 부탁드립니다. 리뷰 요청 주시면 바로 머지하겠습니다!

givxq committed:
[feature] 2step_refactoring
- Console main 프로그램 변경

givxq commented:
좋은 피드백 감사합니다. 
저도 스프링 기능을 사용하지 않으려고도 고민했었는데, 말씀해주신 방법이 도움이 되었습니다🙏🏻
그동안 좋은 코멘트&피드백 감사합니다!

seokhongkim reviewed (approved):
진범님, 이번 코멘트도 잘 반영하셨네요! 👍 
2단계 PR도 머지하겠습니다.
웹과 콘솔 환경에서 모두 실행할 수 있는 자동차 경주 게임 잘 구현하셨고, 미션 진행하느라 수고하셨습니다! 😀
